lambda/lambda.ml

```diff
@@ -673,79 +673,101 @@ let rec make_sequence fn = function
-                        sw_failaction = subst_opt s sw.sw_failaction; },
+        Lswitch(subst s l arg,
+                {sw with sw_consts = List.map (subst_case s l) sw.sw_consts;
+                        sw_blocks = List.map (subst_case s l) sw.sw_blocks;
+                        sw_failaction = subst_opt s l sw.sw_failaction; },
                 loc)
     | Lstringswitch (arg,cases,default,loc) ->
         Lstringswitch
-          (subst s arg,List.map (subst_strcase s) cases,subst_opt s default,loc)
-    | Lstaticraise (i,args) ->  Lstaticraise (i, subst_list s args)
+          (subst s l arg,
+           List.map (subst_strcase s l) cases,
+           subst_opt s l default,loc)
+    | Lstaticraise (i,args) ->  Lstaticraise (i, subst_list s l args)
     | Lstaticcatch(body, (id, params), handler) ->
-        Lstaticcatch(subst s body, (id, params),
-                    subst (remove_list params s) handler)
+        let params, l' = bvs params l in
+        Lstaticcatch(subst s l body, (id, params),
+                    subst s l' handler)
     | Ltrywith(body, exn, handler) ->
-        Ltrywith(subst s body, exn, subst (Ident.Map.remove exn s) handler)
-    | Lifthenelse(e1, e2, e3) -> Lifthenelse(subst s e1, subst s e2, subst s e3)
-    | Lsequence(e1, e2) -> Lsequence(subst s e1, subst s e2)
-    | Lwhile(e1, e2) -> Lwhile(subst s e1, subst s e2)
+        let exn, l' = bv exn l in
+        Ltrywith(subst s l body, exn, subst s l' handler)
+    | Lifthenelse(e1, e2, e3) ->
+       Lifthenelse(subst s l e1, subst s l e2, subst s l e3)
+    | Lsequence(e1, e2) -> Lsequence(subst s l e1, subst s l e2)
+    | Lwhile(e1, e2) -> Lwhile(subst s l e1, subst s l e2)
     | Lfor(v, lo, hi, dir, body) ->
-        Lfor(v, subst s lo, subst s hi, dir,
-          subst (Ident.Map.remove v s) body)
+        let v, l' = bv v l in
+        Lfor(v, subst s l lo, subst s l hi, dir,
+          subst s l' body)
     | Lassign(id, e) ->
-        assert(not (Ident.Map.mem id s));
-        Lassign(id, subst s e)
+       assert (not (Ident.Map.mem id s));
+        let id =
+          try Ident.Map.find id l with Not_found -> id in
+        Lassign(id, subst s l e)
     | Lsend (k, met, obj, args, loc) ->
-        Lsend (k, subst s met, subst s obj, subst_list s args, loc)
+        Lsend (k, subst s l met, subst s l obj, subst_list s l args, loc)
     | Levent (lam, evt) ->
+        let lev_env = evt.lev_env in
+        let lev_env =
+          Ident.Map.fold (fun id id' env ->
+            match Env.find_value (Path.Pident id) evt.lev_env with
+              | exception Not_found -> env
+              | vd -> Env.add_value id' vd env
+            ) l lev_env
+        in
         let lev_env =
           Ident.Map.fold (fun id _ env ->
             match Env.find_value (Path.Pident id) evt.lev_env with
             | exception Not_found -> env
             | vd -> update_env id vd env
-          ) s evt.lev_env
+          ) s lev_env
```

Comment:
@chambart just asked me why there are two separate environments `s` (for free variables of the original term) and `l` (for bound variables traversed so far); the two environments are handled differently here: for free variables substituted by the user, we use the user-provided function `update_env`; for bound variables substituted in `l`, I do the correct thing which is to change the binding to the new bound variable name. (Ideally I would also remove the old binding to the substituted-away variable, but there is no API to remove a binding from debug environments.)